ckanext/opendata_theme/opengov_custom_header/controller.py

```diff
@@ -0,0 +1,184 @@
+# encoding: utf-8
+import ast
+
+import ckan.controllers.admin as admin
+import ckan.lib.navl.dictization_functions as dict_fns
+import six
+from ckan.logic import (
+    clean_dict,
+    tuplize_dict,
+    parse_params, ValidationError
+)
+from ckan.plugins.toolkit import (
+    get_action,
+    redirect_to,
+    render,
+    request
+)
+from webhelpers.html import literal
+
+from ckanext.opendata_theme.opengov_custom_header.constants import CONFIG_SECTION, CONTROLLER, DEFAULT_CONFIG_SECTION
+
+
+class Header(object):
+    def __init__(self, title, link, position, html='', active=False):
+        self.title = six.text_type(title).lower()
+        self.link = six.text_type(link)
+        self.position = position
+        self._html = None
+        self.html = six.text_type(html)
+        self.active = active
+
+    def __repr__(self):
+        return '{}:{}'.format(self.position, self.title)
+
+    def to_dict(self):
+        return {
+            'title': self.title,
+            'link': self.link,
+            'position': self.position,
+            'html': self._html,
+            'active': self.active,
+        }
+
+    @property
+    def html(self):
+        if not self._html:
+            self._html = literal('<li><a href="{}">{title}</a></li>'.format(
+                self.link, title=self.title))
+        return literal(self._html)
+
+    @html.setter
+    def html(self, value):
+        self._html = value
+
+
+class CustomHeaderController(admin.AdminController):
+    def remove_link(self):
+        if request.method == 'POST':
+            header_data = self.get_custom_header_metadata()
+            data = clean_dict(dict_fns.unflatten(
+                tuplize_dict(parse_params(request.POST))))
+            item = [link for link in header_data['links'] if link.title == data['to_remove']]
+            header_data['links'].remove(item[0])
+            error = self.save_header_metadata(header_data)
+            header_data['errors'] = error
+            self.redirect_to_custom_header_page(header_data)
+
+    def add_link(self):
+        if request.method == 'POST':
+            header_data = self.get_custom_header_metadata()
+            data = clean_dict(dict_fns.unflatten(
+                tuplize_dict(parse_params(request.POST))))
+            header_data['links'].append(
+                Header(
+                    title=data.get('new_title'),
+                    link=data.get('new_link'),
+                    position=len(header_data['links'])+1,
+                ))
+            error = self.save_header_metadata(header_data)
+            header_data['errors'] = error
+            return render('admin/custom_header.html',
+                          extra_vars=header_data)
+
+    def custom_header(self):
+        custom_header = self.get_custom_header_metadata()
+        if not custom_header.get('links'):
```

Comment:
When reset_custom_header() is called `links` will be set to an empty list. This results in an infinite loop. Instead this line can be changed to `if not custom_header:`

Reply:
done